Update realtime and mutation package for new txn system

Junoth opened the pull request:
This PR is the first one to update existing methods to adopt new transaction system, which was added in this PR: https://github.com/teamyapp/teamy-backend/pull/363

To avoid affecting other methods, we add v2Dao and v2Mutation packages and according methods for new implementation.

Also, we updated the mutation and passed in client notifiers when creation instead of call DAO in the mthod.

Junoth commented on core/mutationv2/update_task_mutation.go:
```diff
@@ -0,0 +1,69 @@
+
+func NewUpdateTaskMutation(
+	dataCollector telemetry.DataCollector,
+	clientNotifiers []*realtimev2.ClientNotifier,
```
@magicoder10 I have to update the mutation to pass in client notifiers when creation instead of querying when calling the method, since this will be called when realtime transaction commit which is outside of db transaction. Please let me know if you have concerns.

naturetime10 replied:
Certain mutations, such as adding an user to a team, will need to send the transaction to the newly added user. How about we add a new API to mutation `PrepareClientNotifiers(ct context.Context, sqlTx *sql.Tx)` which cache client notifiers into instance variable of the `Mutation`? It will be invoked before calling `sqlTx.Commit()`

naturetime10 commented on core/daov2/sqldb/sprint.go:
```diff
@@ -0,0 +1,282 @@
+	return nil
+}
+
+func NewSprint(dataCollector telemetry.DataCollector, sqlDB *sql.DB) Sprint {
```
```suggestion
func NewSprint(dataCollector telemetry.DataCollector) Sprint {
```

naturetime10 commented on core/daov2/sqldb/sprint_task_relation.go:
```diff
@@ -0,0 +1,161 @@
+	return nil
+}
+
+func NewSprintTaskRelation(dataCollector telemetry.DataCollector, db *sql.DB) SprintTaskRelation {
```
```suggestion
func NewSprintTaskRelation(dataCollector telemetry.DataCollector) SprintTaskRelation {
```

naturetime10 commented on core/mutationv2/update_sprint_participant_mutation.go:
```diff
@@ -0,0 +1,69 @@
+
+func NewUpdateSprintParticipantMutation(
+	dataCollector telemetry.DataCollector,
+	clientNotifiers []*realtimev2.ClientNotifier,
```
Certain mutations, such as adding an user to a team, will need to send the transaction to the newly added user. How about we add a new API to mutation `PrepareClientNotifiers(ct context.Context, sqlTx *sql.Tx)` which cache client notifiers into instance variable of the `Mutation`? It will be invoked before calling `sqlTx.Commit()`

naturetime10 commented on core/service/task.go:
```diff
@@ -293,92 +303,120 @@ func (t Task) UpdateTask(ct context.Context, taskID uint64, input UpdateTaskInpu
+		task.OwningTeamID = input.OwningTeamID
+		task.Effort = input.Effort
+		task.DueAt = input.DueAt
+		updatedAt := time.Now()
```
```suggestion
		updatedAt := time.Now().UTC()
```

naturetime10 commented on core/service/task.go:
```diff
@@ -293,92 +303,120 @@ func (t Task) UpdateTask(ct context.Context, taskID uint64, input UpdateTaskInpu
+		task.UpdatedAt = &updatedAt
+
+		var clientNotifiers []*realtimev2.ClientNotifier
+		clientNotifiers, internalErr = t.stateSyncerV2.GetClientNotifiersByTeamID(ct, task.OwningTeamID)
```
Need to handle error

naturetime10 commented on core/service/task.go:
```diff
@@ -293,92 +303,120 @@ func (t Task) UpdateTask(ct context.Context, taskID uint64, input UpdateTaskInpu
+			t.dataCollector.Logger.ErrorWithContext(ct, internalErr)
+			return internalErr
+		}
+		rtTx.AppendMutation(updateTaskMutation)
```
```suggestion
		}
		
		rtTx.AppendMutation(updateTaskMutation)
```

Junoth committed:
Prepare new methods and deprecate old ones for new transaction system

We found that hard to maintain if we create a new realtime v2 package since there are shared state especially in state_syncer. We decide to reuse the realtime package to add new methods and deprecate old methods

Also, update mutations to add new methods and deprecate old ones

Junoth committed:
Update transaction to use realtime package

naturetime10 commented on core/mutation/create_invitation_mutation.go:
```diff
@@ -18,6 +19,18 @@ type CreateInvitationMutation struct {
 
+var _ realtime.Mutation = (*CreateInvitationMutation)(nil)
+
+func (c *CreateInvitationMutation) ExecuteV2(ct context.Context, tx *sql.Tx) *errs.Error {
```
Let's move `ExecuteV2` and `PrepareClientNotifiers` below `GetID`.

naturetime10 commented on core/mutation/create_message_mutation.go:
```diff
@@ -19,6 +20,18 @@ type CreateMessageMutation struct {
 	message       entity.Message
 }
 
+var _ realtime.Mutation = (*CreateMessageMutation)(nil)
```
Let's move `ExecuteV2` and `PrepareClientNotifiers` below `GetID`.

naturetime10 commented on core/mutation/create_sprint_participant_mutation.go:
```diff
@@ -19,6 +20,18 @@ type CreateSprintParticipantMutation struct {
 
+var _ realtime.Mutation = (*CreateSprintParticipantMutation)(nil)
+
+func (c *CreateSprintParticipantMutation) ExecuteV2(ct context.Context, tx *sql.Tx) *errs.Error {
```
Let's move `ExecuteV2` and `PrepareClientNotifiers` below `GetID`.

naturetime10 commented on core/mutation/create_sprint_task_relation_mutation.go:
```diff
@@ -19,6 +20,18 @@ type CreateSprintTaskRelationMutation struct {
 
+var _ realtime.Mutation = (*CreateSprintTaskRelationMutation)(nil)
+
+func (c *CreateSprintTaskRelationMutation) ExecuteV2(ct context.Context, tx *sql.Tx) *errs.Error {
```
Let's move `ExecuteV2` and `PrepareClientNotifiers` below `GetID`.

naturetime10 commented on core/mutation/create_task_await_for_relation_mutation.go:
```diff
@@ -19,6 +20,18 @@ type CreateTaskAwaitForRelationMutation struct {
 
+var _ realtime.Mutation = (*CreateTaskAwaitForRelationMutation)(nil)
+
+func (c *CreateTaskAwaitForRelationMutation) ExecuteV2(ct context.Context, tx *sql.Tx) *errs.Error {
```
Let's move `ExecuteV2` and `PrepareClientNotifiers` below `GetID`.

naturetime10 commented on core/mutation/create_task_link_mutation.go:
```diff
@@ -19,6 +20,18 @@ type CreateTaskLinkMutation struct {
 
+var _ realtime.Mutation = (*CreateTaskLinkMutation)(nil)
+
+func (c *CreateTaskLinkMutation) ExecuteV2(ct context.Context, tx *sql.Tx) *errs.Error {
```
Let's move `ExecuteV2` and `PrepareClientNotifiers` below `GetID`.

naturetime10 commented on core/mutation/create_task_mutation.go:
```diff
@@ -18,6 +19,18 @@ type CreateTaskMutation struct {
 
+var _ realtime.Mutation = (*CreateTaskMutation)(nil)
+
+func (c *CreateTaskMutation) ExecuteV2(ct context.Context, tx *sql.Tx) *errs.Error {
```
Let's move `ExecuteV2` and `PrepareClientNotifiers` below `GetID`.

naturetime10 commented on core/mutation/create_team_member_muation.go:
```diff
@@ -18,6 +19,18 @@ type CreateTeamMemberMutation struct {
 
+var _ realtime.Mutation = (*CreateTeamMemberMutation)(nil)
+
+func (c *CreateTeamMemberMutation) ExecuteV2(ct context.Context, tx *sql.Tx) *errs.Error {
```
Let's move `ExecuteV2` and `PrepareClientNotifiers` below `GetID`.

naturetime10 commented on core/mutation/create_team_mutation.go:
```diff
@@ -18,6 +19,18 @@ type CreateTeamMutation struct {
 
+var _ realtime.Mutation = (*CreateTeamMutation)(nil)
+
+func (c *CreateTeamMutation) ExecuteV2(ct context.Context, tx *sql.Tx) *errs.Error {
```
Let's move `ExecuteV2` and `PrepareClientNotifiers` below `GetID`.

naturetime10 commented on core/mutation/delete_invitation_mutation.go:
```diff
@@ -18,6 +19,18 @@ type DeleteInvitationMutation struct {
 
+var _ realtime.Mutation = (*DeleteInvitationMutation)(nil)
+
+func (d *DeleteInvitationMutation) ExecuteV2(ct context.Context, tx *sql.Tx) *errs.Error {
```
Let's move `ExecuteV2` and `PrepareClientNotifiers` below `GetID`.

naturetime10 commented on core/mutation/delete_message_mutation.go:
```diff
@@ -19,6 +20,18 @@ type DeleteMessageMutation struct {
 
+var _ realtime.Mutation = (*DeleteMessageMutation)(nil)
+
+func (d *DeleteMessageMutation) ExecuteV2(ct context.Context, tx *sql.Tx) *errs.Error {
```
Let's move `ExecuteV2` and `PrepareClientNotifiers` below `GetID`.

naturetime10 commented on core/mutation/delete_sprint_participant_mutation.go:
```diff
@@ -19,6 +20,18 @@ type DeleteSprintParticipantMutation struct {
 
+var _ realtime.Mutation = (*DeleteSprintParticipantMutation)(nil)
+
+func (d *DeleteSprintParticipantMutation) ExecuteV2(ct context.Context, tx *sql.Tx) *errs.Error {
```
Let's move `ExecuteV2` and `PrepareClientNotifiers` below `GetID`.

naturetime10 commented on core/mutation/delete_sprint_task_relation_mutation.go:
```diff
@@ -19,6 +20,18 @@ type DeleteSprintTaskRelationMutation struct {
 
+var _ realtime.Mutation = (*DeleteSprintTaskRelationMutation)(nil)
+
+func (d *DeleteSprintTaskRelationMutation) ExecuteV2(ct context.Context, tx *sql.Tx) *errs.Error {
```
Let's move `ExecuteV2` and `PrepareClientNotifiers` below `GetID`.

naturetime10 commented on core/mutation/delete_task_await_for_relation_mutation.go:
```diff
@@ -19,6 +20,18 @@ type DeleteTaskAwaitForRelationMutation struct {
 
+var _ realtime.Mutation = (*DeleteTaskAwaitForRelationMutation)(nil)
+
+func (d *DeleteTaskAwaitForRelationMutation) ExecuteV2(ct context.Context, tx *sql.Tx) *errs.Error {
```
Let's move `ExecuteV2` and `PrepareClientNotifiers` below `GetID`.

naturetime10 commented on core/mutation/delete_task_mutation.go:
```diff
@@ -18,6 +19,18 @@ type DeleteTaskMutation struct {
 
+var _ realtime.Mutation = (*DeleteTaskMutation)(nil)
+
+func (d *DeleteTaskMutation) ExecuteV2(ct context.Context, tx *sql.Tx) *errs.Error {
```
Let's move `ExecuteV2` and `PrepareClientNotifiers` below `GetID`.

naturetime10 commented on core/mutation/delete_team_member_muation.go:
```diff
@@ -19,6 +20,18 @@ type DeleteTeamMemberMutation struct {
+	panic("implement me")
+}
+
+func (d *DeleteTeamMemberMutation) PrepareClientNotifiers(ct context.Context, tx *sql.Tx) ([]*realtime.ClientNotifier, *errs.Error) {
```
Let's move `ExecuteV2` and `PrepareClientNotifiers` below `GetID`.

naturetime10 commented on core/mutation/delete_team_mutation.go:
```diff
@@ -19,6 +20,16 @@ type DeleteTeamMutation struct {
+	panic("implement me")
+}
+
+func (u *DeleteTeamMutation) PrepareClientNotifiers(ct context.Context, tx *sql.Tx) ([]*realtime.ClientNotifier, *errs.Error) {
```
Let's move `ExecuteV2` and `PrepareClientNotifiers` below `GetID`.

naturetime10 commented on core/mutation/update_invitation_mutation.go:
```diff
@@ -18,8 +19,20 @@ type UpdateInvitationMutation struct {
-	return d.id
+var _ realtime.Mutation = (*UpdateInvitationMutation)(nil)
+
+func (u *UpdateInvitationMutation) ExecuteV2(ct context.Context, tx *sql.Tx) *errs.Error {
```
Let's move `ExecuteV2` and `PrepareClientNotifiers` below `GetID`.

naturetime10 commented on core/mutation/update_message_mutation.go:
```diff
@@ -19,6 +20,18 @@ type UpdateMessageMutation struct {
 
+var _ realtime.Mutation = (*UpdateMessageMutation)(nil)
+
+func (u *UpdateMessageMutation) ExecuteV2(ct context.Context, tx *sql.Tx) *errs.Error {
```
Let's move `ExecuteV2` and `PrepareClientNotifiers` below `GetID`.

naturetime10 commented on core/mutation/update_task_activity_mutation.go:
```diff
@@ -18,6 +19,18 @@ type UpdateTaskActivityMutation struct {
 
+var _ realtime.Mutation = (*UpdateTaskActivityMutation)(nil)
+
+func (u *UpdateTaskActivityMutation) ExecuteV2(ct context.Context, tx *sql.Tx) *errs.Error {
```
Let's move `ExecuteV2` and `PrepareClientNotifiers` below `GetID`.

naturetime10 commented on core/mutation/update_task_mutation.go:
```diff
@@ -18,6 +19,18 @@ type UpdateTaskMutation struct {
 
+var _ realtime.Mutation = (*UpdateTaskMutation)(nil)
+
+func (u *UpdateTaskMutation) ExecuteV2(ct context.Context, tx *sql.Tx) *errs.Error {
```
Let's move `ExecuteV2` and `PrepareClientNotifiers` below `GetID`.

naturetime10 commented on core/mutation/update_team_member_muation.go:
```diff
@@ -18,6 +19,18 @@ type UpdateTeamMemberMutation struct {
 
+var _ realtime.Mutation = (*UpdateTeamMemberMutation)(nil)
+
+func (u *UpdateTeamMemberMutation) ExecuteV2(ct context.Context, tx *sql.Tx) *errs.Error {
```
Let's move `ExecuteV2` and `PrepareClientNotifiers` below `GetID`.

naturetime10 commented on core/mutation/update_team_mutation.go:
```diff
@@ -18,6 +19,18 @@ type UpdateTeamMutation struct {
 
+var _ realtime.Mutation = (*UpdateTeamMutation)(nil)
+
+func (u *UpdateTeamMutation) ExecuteV2(ct context.Context, tx *sql.Tx) *errs.Error {
```
Let's move `ExecuteV2` and `PrepareClientNotifiers` below `GetID`.

naturetime10 commented on core/mutation/update_user_mutation.go:
```diff
@@ -19,6 +20,18 @@ type UpdateUserMutation struct {
 
+var _ realtime.Mutation = (*UpdateUserMutation)(nil)
+
+func (u *UpdateUserMutation) ExecuteV2(ct context.Context, tx *sql.Tx) *errs.Error {
```
Let's move `ExecuteV2` and `PrepareClientNotifiers` below `GetID`.

naturetime10 commented on core/realtime/mutation.go:
```diff
@@ -16,9 +17,13 @@ const (
 	CleanUp(ct context.Context) *errs.Error
+	// Deprecated: The old method should not be used anymore. Use another PrepareClientNotifiers method instead
 	GetClientNotifiers(ct context.Context) ([]*ClientNotifier, *errs.Error)
+	PrepareClientNotifiers(ct context.Context, tx *sql.Tx) ([]*ClientNotifier, *errs.Error)
```
```suggestion
	PrepareClientNotifiers(ct context.Context, tx *sql.Tx) *errs.Error
```

We are going to change internal state of mutation instead of returning `[]*ClientNotifier`

naturetime10 commented on core/realtime/transaction.go:
```diff
@@ -78,6 +84,12 @@ func (t *Transaction) ApplyMutation(ct context.Context, mutation Mutation) *errs
+func (t *Transaction) AppendMutation(mutation Mutation) *errs.Error {
+	t.mutations = append(t.mutations, mutation)
+	return nil
+}
```
Let's add a method called `GetMutations() []Mutation`.

naturetime10 commented on core/service/transaction.go:
```diff
@@ -30,7 +30,7 @@ func (t TransactionsContext) withTransactions(execute func(sqlTx *sql.Tx, rtTx *
 	}
 
-	rtTx := realtimev2.NewTransaction(t.dataCollector, t.stateSyncer)
+	rtTx := realtime.NewTransaction(t.dataCollector, t.stateSyncer)
```
Please get and loop through all mutations from rtTx and invoke prepare client notifiers between line 34 and line 35.

naturetime10 reviewed (changes requested):
It seems that we are not including updateTask changes in TaskService.

Junoth committed:
Update according to comment
